hs/app/reach/Main.hs

```diff
@@ -0,0 +1,351 @@
+{-# OPTIONS_GHC -fno-warn-type-defaults #-}
+
+{-# LANGUAGE ExtendedDefaultRules #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE QuasiQuotes          #-}
+
+module Main (main) where
+
+import Control.Monad (join)
+import Control.Monad.Shell
+import Options.Applicative
+import Options.Applicative.Help.Pretty
+import System.Posix.IO (stdError)
+
+import qualified Data.Text         as TS
+import qualified Data.Text.Lazy    as TL
+import qualified Data.Text.Lazy.IO as TL
+import qualified NeatInterpolation as N
+
+import qualified Reach.Version     as V
+
+default (TL.Text)
+
+
+type Subcommand = ParserInfo (Script ())
+
+
+reachImages :: [TL.Text]
+reachImages =
+  [ "reach"
+  , "ethereum-devnet"
+  , "algorand-devnet"
+  , "devnet-cfx"
+  , "runner"
+  , "react-runner"
+  , "rpc-server"
+  ]
+
+
+--------------------------------------------------------------------------------
+echo :: CmdParams p => p
+echo = cmd "echo"
+
+
+exit :: Int -> Script ()
+exit i = cmd "exit" $ static i
+
+
+docker :: CmdParams p => p
+docker = cmd "docker"
+
+
+rm :: CmdParams p => p
+rm = cmd "rm"
+
+
+stdErrDevNull :: Script () -> Script ()
+stdErrDevNull f = f |> (stdError, TL.unpack "/dev/null")
+
+
+-- | Squelch @stderr@ and continue even if @f@ returns non-zero exit code
+regardless :: Script () -> Script ()
+regardless f = stdErrDevNull f -||- cmd ":"
+
+
+-- | A completely silent 'regardless'
+regardless' :: Script () -> Script ()
+regardless' f = regardless $ toStderr f
+
+
+--------------------------------------------------------------------------------
+clean :: Subcommand
+clean = info f $ fullDesc <> desc <> fdoc where
+  desc = progDesc "Delete 'build/$MODULE.$IDENT.mjs'"
+  fdoc = footerDoc . Just
+     $  text "MODULE is \"index\" by default"
+   <$$> text "IDENT  is \"main\"  by default"
+   <$$> text ""
+   <$$> text "If:"
+   <$$> text " * MODULE is a directory then `cd $MODULE && rm -f \"build/index.$IDENT.mjs\";"
+   <$$> text " * MODULE is <something-else> then `rm -f \"build/$MODULE.$IDENT.mjs\""
+
+  go m i = do
+    let f' m' = rm "-f" $ "build/" <> TL.pack m' <> "." <> i <> ".mjs"
+
+    case m of
+      "index" -> f' m
+      _       -> ifCmd (test $ TDirExists m)
+        (cmd "cd" m -||- exit 1 *> f' "index")
+        (f' m)
+
+  f = go
+    <$> strArgument (metavar "MODULE" <> value "index" <> showDefault)
+    <*> strArgument (metavar "IDENT"  <> value "main"  <> showDefault)
+
+
+--------------------------------------------------------------------------------
+compile :: Subcommand
+compile = info f d where
+  d = progDesc "Compile an app"
+  f = undefined
+
+
+--------------------------------------------------------------------------------
+reachVersionShort :: TS.Text -- TODO
+reachVersionShort = TS.pack V.compatibleVersionStr
+
+
+initRsh :: TS.Text -> TL.Text
+initRsh v = TL.fromStrict [N.text|
```

Comment:
I think these things should be external files that get embedded and/or are in the environment of the Docker image